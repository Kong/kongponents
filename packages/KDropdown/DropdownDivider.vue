<template>
  <li
    :aria-label="label ? label : null"
    role="separator">
    <div>{{ label }}</div>
    <hr>
  </li>
</template>

<script lang="ts">
export default {
  name: 'DropdownDivider',
  props: {
    label: {
      type: String,
      default: '',
      required: false
    }
  }
}
</script>

<style lang="scss" scoped>
li {
  cursor: default;
  color: var(--black-70);
  padding: 0.325rem 1rem 0.65rem 1rem;
  font-weight: 500;
  font-size: 1rem;
  display: flex;
  align-items: center;

  &[role="separator"]:not(:first-child) {
    margin-top: 0.8rem;
  }

  div {
    flex: none;
  }
}

hr {
  border: none;
  /* IE uses color rather than background-color */
  color: var(--black-25);
  background-color: var(--black-25);
  margin-left: 0.4rem;
  width: 100%;
  height: 1px;
}
</style>
