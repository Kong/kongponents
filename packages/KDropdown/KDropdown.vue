<template>
  <div
    :class="{'is-open': isOpen }"
    class="k-dropdown">
    <!-- Overlay to click away -->
    <div
      v-if="isOpen"
      class="overlay"
      @click="isOpen = !isOpen"/>
    <KButton
      :appearance="buttonAppearance"
      :disabled="disabled"
      aria-haspopup="true"
      class="k-dropdown-toggle"
      @click="() => isOpen = !isOpen">
      <span>{{ toggleText }}</span>
    </KButton>
    <ul
      :class="`k-dropdown-menu-${position}`"
      class="k-dropdown-menu"
      role="listbox"
      aria-labelledby="dropdown">
      <li
        v-for="item in items"
        :key="getItem(item)"
        :id="getItem(item)"
        :class="{ 'is-selected': item.selected }"
        class="k-dropdown-item"
        role="option"
        @click="clickHandler(item)">
        <slot
          :name="item"
          :item="item">{{ getItem(item) }}</slot>
      </li>
    </ul>
  </div>
</template>

<script>
import { default as KButton, appearances } from '@kongponents/kbutton/KButton.vue'

export const alignments = {
  left: 'left',
  center: 'center',
  right: 'right'
}

export default {
  name: 'KDropdown',
  components: { KButton },

  props: {
    /**
     * Align dropdown menu to left, center or right
     */
    position: {
      type: String,
      default: 'left',
      validator: (value) => Object.values(alignments).includes(value)
    },
    /**
     * Button text
     */
    toggleText: {
      type: String,
      default: ''
    },
    /**
     * Style as button or text link
     */
    buttonAppearance: {
      type: String,
      default: 'secondary',
      validator: (value) => Object.values(appearances).includes(value)
    },
    /**
     * Array of items to render in menu. Can pass array of Strings or Objects
     * String: ['item', 'item', 'item]
     * Object: [ { name: 'item', selected: false }, { name: 'item'} ]
     */
    items: {
      type: Array,
      default: () => []
    },
    /**
     * Disable dropdown toggle button
     */
    disabled: {
      type: Boolean,
      default: false
    }
  },

  data () {
    return {
      isOpen: false
    }
  },

  methods: {
    getItem (item) {
      return (item && item.name) || item
    },

    clickHandler (item) {
      // Only run if item has selected key
      if (Object.keys(item).includes('selected')) {
        const itemIndex = this.items.findIndex(i => i.name === item.name)

        item.selected = !item.selected
        this.$emit('selected', item, itemIndex)
      }
    }
  }
}
</script>

<style scoped lang="scss">
@import '~@kongponents/styles/_variables.scss';

.overlay {
  position: fixed;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
}

.k-dropdown {
  position: relative;
  display: inline-block;
  .k-dropdown-toggle:after {
    display: inline-block;
    width: 0;
    height: 0;
    margin-left: var(--spacing-xs);
    vertical-align: middle;
    content: "";
    border-top: 0.325em solid;
    border-right: 0.325em solid transparent;
    border-left: 0.325em solid transparent;
  }
  &.is-open .k-dropdown-toggle:after {
    transform: rotate(180deg);
  }
}

/* Dropdown menu */
.k-dropdown.is-open .k-dropdown-menu {
  display: block;
}
.k-dropdown .k-dropdown-menu {
  display: none;
  position: absolute;
  right: 0;
  left: auto;
  min-width: 100px;
  max-width: 300px;
  width: auto;
  margin: 0;
  padding: var(--spacing-xs) 0;
  list-style: none;
  border-radius: 3px;
  background: #fff;
  border: 1px solid var(--grey-88);
  box-shadow: 0 2px 12px rgba(0, 0, 0, 0.15);
  z-index: 9999;
}
/* Dropdown menu alignment */
.k-dropdown .k-dropdown-menu-left {
  left: 0;
  right: auto;
}
.k-dropdown .k-dropdown-menu-right {
  left: auto;
  right: 0;
}
.k-dropdown .k-dropdown-menu-center {
  left: 50%;
  right: auto;
  transform: translateX(-50%);
  -webkit-transform: translateX(-50%);
}

/* Menu items */
.k-dropdown .k-dropdown-menu .k-dropdown-item {
  display: block;
  padding: 0 var(--spacing-md);
  white-space: nowrap;
  color: var(--tblack-70);
  line-height: 2;
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
  transition: all 200ms ease-in;
}
.k-dropdown .k-dropdown-menu .k-dropdown-item.is-selected {
  background-image: url("data:image/svg+xml,%3Csvg xmlns='http://www.w3.org/2000/svg' width='16' height='13'%3E%3Cpath fill='%2366BFFF' fill-rule='evenodd' d='M4.5 10L13.7968194.7031806c.3883559-.3883559 1.021292-.3850692 1.4092576.0028964l.087846.087846c.3899555.3899555.3864986 1.0256554.0040118 1.4081422L5.2020652 12.2979348c-.3877399.3877399-1.0256941.3784363-1.4051902-.0010598l-3.09375-3.09375C.3147998 8.8147998.3181114 8.1818886.706077 7.793923l.087846-.087846c.3899555-.3899555 1.0271296-.3850244 1.4170351.0048811L4.5 10z'/%3E%3C/svg%3E");
  background-position: var(--spacing-sm) 50%;
  background-repeat: no-repeat;
  padding-left: var(--spacing-xl);
  color: var(--blue-link);
}
.k-dropdown .k-dropdown-menu .k-dropdown-item:hover,
.k-dropdown .k-dropdown-menu .k-dropdown-item.active {
  // background-color: var(--blue-lightest);
  // cursor: pointer;
  // transition: all 200ms ease-out;
}
.k-dropdown .k-dropdown-menu .k-dropdown-item a,
.k-dropdown .k-dropdown-menu .k-dropdown-item > div > a {
  // color: inherit;
  // display: block;
}
</style>

<style>
.k-dropdown .k-dropdown-toggle:focus,
.k-dropdown .k-dropdown-toggle:active {
  outline: 0;
}

</style>
