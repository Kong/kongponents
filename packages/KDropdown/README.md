# @kongponents/kdropdown

[![](https://img.shields.io/npm/v/@kongponents/kdropdown.svg?style=flat-square)](https://www.npmjs.com/package/@kongponents/kdropdown)

```vue
<KDropdown
  toggle-text="Items"
  :items="['one', 'two', 'three']"/>
```
