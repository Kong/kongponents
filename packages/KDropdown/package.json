{
  "name": "@kongponents/kdropdown",
  "version": "0.1.0",
  "description": "Simple",
  "main": "dist/KDropdown.umd.min.js",
  "componentName": "KDropdown",
  "source": "KDropdown.vue",
  "files": [
    "dist",
    "KDropdown.vue"
  ],
  "repository": {
    "type": "git",
    "url": "git+https://github.com/Kong/kongponents.git"
  },
  "author": "Kong Inc",
  "license": "Apache-2.0",
  "bugs": {
    "url": "https://github.com/Kong/kongponents/issues"
  },
  "homepage": "https://github.com/Kong/kongponents#readme",
  "publishConfig": {
    "access": "public"
  },
  "dependencies": {
    "@kongponents/kbutton": "^0.1.16"
  }
}
