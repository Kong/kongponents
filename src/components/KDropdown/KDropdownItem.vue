<template>
  <li
    class="k-dropdown-item"
    :class="{
      'has-divider': hasDivider,
      'disabled': disabled,
      'danger': danger || isDangerous,
      'dropdown-selected-option': selected
    }"
    data-testid="dropdown-item"
  >
    <component
      :is="availableComponents[componentType].tag"
      data-testid="dropdown-item-trigger"
      v-bind="availableComponents[componentType].attrs"
      @click="availableComponents[componentType].onClick"
    >
      <span
        class="dropdown-item-trigger-label"
      >
        <slot>{{ label }}</slot>
      </span>
    </component>
  </li>
</template>

<script lang="ts" setup>
import type { DropdownItem, DropdownItemRenderedRecord, DropdownItemRenderedType, DropdownItemType } from '@/types'
import type { PropType } from 'vue'
import { computed, useAttrs } from 'vue'

const attrs = useAttrs()

const props = defineProps({
  item: {
    type: Object as PropType<DropdownItem>,
    default: null,
    // Items must have a label
    validator: (item: DropdownItem) => item.label !== undefined,
  },
  /**
   * Use this prop to add a divider above the item.
   */
  hasDivider: {
    type: Boolean,
    default: false,
  },
  danger: {
    type: Boolean,
    default: false,
  },
  disabled: {
    type: Boolean,
    default: false,
  },
  selected: {
    type: Boolean,
    default: false,
  },
  /**
   * Internal use only - for tracking selection in conjunction with items prop.
   */
  selectionMenuChild: {
    type: Boolean,
    default: false,
  },
  onClick: {
    type: Function,
    default: undefined,
  },
  /**
   * @deprecated in favor of `danger`
   */
  isDangerous: {
    type: Boolean,
    default: false,
    validator: (value: boolean): boolean => {
      if (value) {
        console.warn('KDropdownItem: `isDangerous` prop is deprecated. Please use `danger` prop instead. See the migration guide for more details: https://alpha--kongponents.netlify.app/guide/migrating-to-version-9.html#kdropdownmenu')
      }

      return true
    },
  },
})

const emit = defineEmits<{
  (e: 'click', val: Event): void;
  (e: 'change', item: DropdownItem): void;
}>()

const type = computed((): DropdownItemType => {
  if (props.item?.to) {
    return 'link'
  } else if (typeof props.onClick !== 'undefined' || props.selectionMenuChild) {
    // checking attrs since we deleted click from listeners
    return 'button'
  }

  return 'default'
})

const label = computed((): string => (props.item?.label) || '')

const to = computed((): string | object | undefined => (props.item?.to) || undefined)

const handleClick = (event: Event): void => {
  emit('click', event)

  if (props.selectionMenuChild) {
    emit('change', props.item)
  }
}

const componentType = computed((): DropdownItemRenderedType => {
  let result: DropdownItemRenderedType = 'div'

  if (type.value === 'link' && to.value && typeof to.value === 'string') {
    result = 'link'
  } else if (type.value === 'link' && to.value && typeof to.value === 'object') {
    result = 'router-link'
  } else if (type.value === 'button') {
    result = 'button'
  }

  return result
})

// Strips falsy `disabled` attribute, so it does not fall onto native <a> elements.
const strippedAttrs = computed((): typeof attrs => {
  const modifiedAttrs = Object.assign({}, attrs)

  // remove some attributes because we want them on wrapper element not the trigger element
  delete modifiedAttrs.class
  delete modifiedAttrs.disabled
  delete modifiedAttrs.style
  // Ensure the `data-testid` attribute is only applied to the top-most element
  delete modifiedAttrs['data-testid']

  return modifiedAttrs
})

// do not change this value, it's used in tests
const dropdownItemTriggerClass = 'dropdown-item-trigger'

const availableComponents = computed((): DropdownItemRenderedRecord => ({
  link: {
    tag: 'a',
    attrs: {
      // add disabled class instead of disabled attribute because disabled is not a valid attribute for links
      class: `${dropdownItemTriggerClass} ${props.disabled ? 'disabled' : ''}`,
      href: to.value as string,
      ...strippedAttrs.value,
    },
  },
  'router-link': {
    tag: 'router-link',
    onClick: handleClick,
    attrs: {
      // add disabled class instead of disabled attribute because disabled is not a valid attribute for links
      class: `${dropdownItemTriggerClass} ${props.disabled ? 'disabled' : ''}`,
      to: to.value,
      ...strippedAttrs.value,
    },
  },
  button: {
    tag: 'button',
    onClick: handleClick,
    attrs: {
      // don't add disabled class because we want the disabled attribute on the button
      class: dropdownItemTriggerClass,
      disabled: props.disabled,
      ...strippedAttrs.value,
    },
  },
  div: {
    tag: 'div',
    attrs: {
      class: dropdownItemTriggerClass,
      ...strippedAttrs.value,
    },
  },
}))
</script>

<style lang="scss" scoped>
/* Component mixins */

@mixin kDropdownItemTriggerDefaults {
  transition: background-color $kongponentsTransitionDurTimingFunc, color $kongponentsTransitionDurTimingFunc;
}

/* Component styles */

.k-dropdown-item, :deep(.k-dropdown-item) {
  display: flex;
  list-style: none;

  &.has-divider {
    display: flex;
    flex-direction: column;

    &::before {
      border-top: var(--kui-border-width-10, $kui-border-width-10) solid var(--kui-color-border, $kui-color-border);
      content: '';
      margin: var(--kui-space-20, $kui-space-20) 0;
      width: 100%;
    }
  }

  &.danger {
    .dropdown-item-trigger {
      color: var(--kui-color-text-danger, $kui-color-text-danger);

      &:hover:not(:disabled):not(.disabled):not(:focus):not(:active) {
        background-color: var(--kui-color-background-danger-weakest, $kui-color-background-danger-weakest);
        color: var(--kui-color-text-danger, $kui-color-text-danger);
      }

      &:focus:not(:disabled):not(.disabled), &:active:not(:disabled):not(.disabled) {
        background-color: var(--kui-color-background-danger-weaker, $kui-color-background-danger-weaker);
        color: var(--kui-color-text-danger, $kui-color-text-danger);
      }
    }
  }

  &.disabled {
    // add cursor pointer on wrapper element because <a> tag has pointer-events: none;
    cursor: not-allowed;
  }

  &.dropdown-selected-option {
    .dropdown-item-trigger {
      background-color: var(--kui-color-background-primary-weakest, $kui-color-background-primary-weakest);
      color: var(--kui-color-text-primary-stronger, $kui-color-text-primary-stronger);

      &:hover:not(:disabled):not(.disabled):not(:focus):not(:active),
      &:focus:not(:disabled):not(.disabled),
      &:active:not(:disabled):not(.disabled) {
        background-color: var(--kui-color-background-primary-weakest, $kui-color-background-primary-weakest);
      }

      &:disabled, &[disabled], &.disabled {
        background-color: var(--kui-color-background-disabled, $kui-color-background-disabled);
      }
    }
  }

  .dropdown-item-trigger {
    background-color: var(--kui-color-background, $kui-color-background);
    border: 0;
    color: var(--kui-color-text-neutral, $kui-color-text);
    cursor: pointer;
    display: flex;
    flex: 1;
    font-family: var(--kui-font-family-text, $kui-font-family-text);
    font-size: var(--kui-font-size-30, $kui-font-size-30);
    font-weight: var(--kui-font-weight-medium, $kui-font-weight-medium);
    line-height: var(--kui-line-height-40, $kui-line-height-40);
    max-width: 100%;
    padding: var(--kui-space-50, $kui-space-50) var(--kui-space-60, $kui-space-60);
    text-align: left;
    text-decoration: none;
    transition: background-color $kongponentsTransitionDurTimingFunc, color $kongponentsTransitionDurTimingFunc;

    &:focus, &:active, &:focus-visible {
      outline: none;
    }

    &:focus-visible {
      box-shadow: var(--kui-shadow-focus, $kui-shadow-focus);
      z-index: 1; // need this to prevent box shadow being cut off by the next/previous sibling
    }

    &:hover:not(:disabled):not(.disabled):not(:focus):not(:active) {
      background-color: var(--kui-color-background-neutral-weaker, $kui-color-background-neutral-weaker);
    }

    &:focus:not(:disabled):not(.disabled), &:active:not(:disabled):not(.disabled) {
      background-color: var(--kui-color-background-neutral-weak, $kui-color-background-neutral-weak);
    }

    &:disabled, &[disabled], &.disabled {
      color: var(--kui-color-text-disabled, $kui-color-text-disabled);
      cursor: not-allowed;
    }

    // remove pointer events from only <a>
    &.disabled {
      pointer-events: none;
    }

    .dropdown-item-trigger-label {
      align-items: center;
      display: inline-flex;
      gap: var(--kui-space-40, $kui-space-40);
      width: 100%;
    }
  }
}
</style>

<style lang="scss">
// must stay unscoped as it's causing issues with the slotted/nested elements
// all rules must be very specific to avoid conflicts with other components

.k-dropdown-item {
  .dropdown-item-trigger {
    .dropdown-item-trigger-label {
      #{$kongponentsKongIconSelector} {
        height: var(--kui-icon-size-40, $kui-icon-size-40) !important;
        width: var(--kui-icon-size-40, $kui-icon-size-40) !important;
      }

      [role="button"] {
        &:not([disabled]) {
          cursor: pointer;

          &:focus, &:active {
            outline: none;
          }

          &:hover, &:focus {
            color: var(--kui-color-text-neutral-stronger, $kui-color-text-neutral-stronger) !important;
          }
        }

        &[disabled] {
          color: var(--kui-color-text-disabled, $kui-color-text-disabled) !important;
          pointer-events: none;
        }
      }
    }
  }
}
</style>
