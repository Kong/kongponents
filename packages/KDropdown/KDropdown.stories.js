import Vue from 'vue'
import { storiesOf } from '@storybook/vue'
import StoryRouter from 'storybook-vue-router'

import KDropdown from './KDropdown.vue'

storiesOf('KDropdown', module)
  .addDecorator(StoryRouter())
  .add('Default', () => ({
    components: { KDropdown },
    template: `
    <div style="margin-left: 50px">
      <KDropdown
        :items="['item 1', 'item 2', 'item 3']"
        toggle-text="Open Dropdown" />
    </div>
  `}))
  .add('Show Active', () => ({
    components: { KDropdown },
    template: `
    <div style="margin-left: 50px">
      <KDropdown
        :items="['workspaces', 'vitals']"
        show-active
        toggle-text="workspaces">
        <template slot="workspaces">
          <router-link to="/overview">Workspaces</router-link>
        </template>
      </KDropdown>
    </div>
  `}))
  .add('Center Aligned', () => ({
    components: { KDropdown },
    template: `
    <div style="margin-left: 50px">
      <KDropdown
        :items="['item 1', 'item 2', 'item 3']"
        position="center"
        toggle-text="Open Dropdown" />
    </div>
  `}))
  .add('Right Aligned', () => ({
    components: { KDropdown },
    template: `
    <div style="margin-left: 50px">
      <KDropdown
        :items="['item 1', 'item 2', 'item 3']"
        position="right"
        toggle-text="Open Dropdown"/>
    </div>
  `}))
