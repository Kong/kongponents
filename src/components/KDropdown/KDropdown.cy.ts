import { mount } from 'cypress/vue'
import { h } from 'vue'
import KDropdown from '@/components/KDropdown/KDropdown.vue'
import KDropdownItem from '@/components/KDropdown/KDropdownItem.vue'

const defaultMenuItems = [
  { label: 'Props' },
  { label: 'Slots' },
  { label: 'Top' },
]

const selectionMenuItems = [{
  label: 'US (United States)',
  value: 'us',
},
{
  label: 'FR (France)',
  value: 'fr',
}]

describe('KDropdown', () => {
  it('renders props when passed', () => {
    const triggerTextProp = 'Drop it!'

    mount(KDropdown, {
      props: {
        triggerText: triggerTextProp,
        items: defaultMenuItems,
      },
    })

    const triggerBtn = cy.getTestId('dropdown-trigger-button')

    triggerBtn.should('contain.text', triggerTextProp)
    triggerBtn.click()

    cy.getTestId('dropdown-list').should('exist')
    cy.getTestId('dropdown-list').should('be.visible')
    cy.getTestId(`k-dropdown-item-${defaultMenuItems[0].label}`).should('exist')
    cy.getTestId(`k-dropdown-item-${defaultMenuItems[0].label}`).should('contain.text', defaultMenuItems[0].label)
    cy.getTestId(`k-dropdown-item-${defaultMenuItems[1].label}`).should('exist')
    cy.getTestId(`k-dropdown-item-${defaultMenuItems[1].label}`).should('contain.text', defaultMenuItems[1].label)
  })

  it('renders with correct px width', async () => {
    const width = 350

    mount(KDropdown, {
      props: {
        width: width + '',
        items: defaultMenuItems,
      },
    })

    const triggerBtn = cy.getTestId('dropdown-trigger')
    triggerBtn.click()

    cy.get('.k-dropdown-popover').invoke('width').should('eq', width)
  })

  it('renders disabled props when passed', () => {
    const tooltipText = 'A sweet tooltip'

    mount(KDropdown, {
      props: {
        triggerText: 'Click me',
        disabled: true,
        disabledTooltip: tooltipText,
        items: defaultMenuItems,
      },
    })

    // button disabled
    const triggerBtn = cy.getTestId('dropdown-trigger')
    // hover
    triggerBtn.trigger('mouseenter')

    cy.get('.k-tooltip').should('contain.text', tooltipText)
  })

  it('renders correctly when isSelectionMenu', () => {
    mount(KDropdown, {
      props: {
        isSelectionMenu: true,
        items: selectionMenuItems,
      },
    })

    cy.get('.selection-dropdown-menu').should('exist')
  })

  it('renders with selected item', () => {
    const selectedLabel = 'Label 1'

    mount(KDropdown, {
      props: {
        triggerText: 'Click me',
        items: [
          { label: selectedLabel, value: 'label1', selected: true },
          ...selectionMenuItems,
        ],
      },
    })

    const triggerBtn = cy.getTestId('dropdown-trigger')
    triggerBtn.click()
    cy.getTestId('dropdown-list').should('be.visible')

    cy.get('.dropdown-selected-option').should('exist')
    cy.get('.dropdown-selected-option').should('contain.text', selectedLabel)
  })

  it('allows slotting content', async () => {
    const itemSlotContent = 'I am slotted baby!'
    const triggerSlotContent = 'Click Me!'

    mount(KDropdown, {
      slots: {
        items: h('span', {}, itemSlotContent),
        default: h('button', {}, triggerSlotContent),
      },
    })

    const triggerBtn = cy.getTestId('dropdown-trigger')
    triggerBtn.click()
    cy.getTestId('dropdown-list').should('be.visible')

    triggerBtn.should('contain.html', triggerSlotContent)
    cy.get('.k-dropdown-popover').should('contain.html', itemSlotContent)
  })

  it('correctly renders dividers on all item types', () => {
    const itemSlotContent = `
    <KDropdownItem
      @click="() => {}"
      data-testid="button"
    >
      A button
    </KDropdownItem>
    <KDropdownItem
      disabled
      @click="() => {}"
      data-testid="disabled-button"
    >
      Disabled button
    </KDropdownItem>
    <KDropdownItem
      :item="{ label: 'You are here 2', to: { path: '/' } }"
      has-divider
      @click="() => {}"
      data-testid="router-link"
    >
      Router link
    </KDropdownItem>
    <KDropdownItem
      :item="{ label: 'You are here 2', to: { path: '/' } }"
      disabled
      @click="() => {}"
      data-testid="disabled-router-link"
    >
      Disabled router link
    </KDropdownItem>
    <KDropdownItem
      has-divider
      :item="{ label: 'You are here 3', to: 'https://kongponents.konghq.com/' }"
      rel="noopener"
      target="_blank"
      data-testid="external-link"
    >
      External link
    </KDropdownItem>
    <KDropdownItem
      :item="{ label: 'You are here 3', to: 'https://kongponents.konghq.com/' }"
      rel="noopener"
      target="_blank"
      disabled
      data-testid="disabled-external-link"
    >
      Disabled external link
    </KDropdownItem>`

    mount(KDropdown, {
      props: {
        triggerText: 'Click me',
        class: 'test-dropdown',
      },
      slots: {
        items: itemSlotContent,
        default: h('button', {}, 'hello'),
      },
      global: {
        components: {
          KDropdownItem,
        },
      },
    })

    const triggerBtn = cy.getTestId('dropdown-trigger')
    triggerBtn.click()
    cy.getTestId('dropdown-list').should('be.visible')

    cy.getTestId('dropdown-list').eq(0).find('.k-dropdown-item').should('have.length', 6)
    cy.getTestId('dropdown-list').eq(0).find('.has-divider').should('have.length', 2)

    cy.get('button[data-testid="button"]').should('be.visible')
    cy.get('button[data-testid="disabled-button"]').should('be.visible').should('be.disabled')

    cy.get('router-link[data-testid="router-link"]').should('be.visible')
    cy.get('router-link[data-testid="disabled-router-link"]').should('be.visible').should('have.attr', 'disabled')

    cy.get('a[data-testid="external-link"]').should('be.visible')
    cy.get('a[data-testid="disabled-external-link"]').should('be.visible').should('have.attr', 'disabled')
  })
})
