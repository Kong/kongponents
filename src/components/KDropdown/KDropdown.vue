<template>
  <div
    class="k-dropdown"
    :class="{ 'selection-dropdown-menu': selectionMenu }"
  >
    <KToggle v-slot="{ toggle, isToggled }">
      <KPop
        v-bind="boundKPopAttributes"
        data-testid="k-dropdown-popover"
        :hide-popover="hidePopover"
        :on-popover-click="() => handleTriggerToggle(isToggled, toggle, false)"
        @closed="() => handleTriggerToggle(isToggled, toggle, false)"
        @opened="() => handleTriggerToggle(isToggled, toggle, true)"
      >
        <component
          :is="tooltipComponent"
          class="dropdown-trigger"
          data-testid="dropdown-trigger"
          :label="disabledTooltip"
          :max-width="!!disabledTooltip ? '240' : undefined"
          :position="!!disabledTooltip ? 'bottom' : undefined"
          :position-fixed="!!disabledTooltip ? true : undefined"
        >
          <slot
            :is-open="isToggled.value"
            name="default"
          >
            <KButton
              v-if="triggerButtonText || icon"
              :appearance="appearance"
              class="dropdown-trigger-button"
              data-testid="dropdown-trigger-button"
              :disabled="disabled"
              :icon="icon"
            >
              {{ triggerButtonText }}
              <ChevronDownIcon
                v-if="showCaret"
                decorative
              />
            </KButton>
          </slot>
        </component>
        <template #content>
          <ul
            class="dropdown-list"
            data-testid="dropdown-list"
          >
            <slot
              :close-dropdown="handleCloseDropdown"
              :handle-selection="handleSelection"
              :items="items"
              name="items"
            >
              <KDropdownItem
                v-for="(item, idx) in items"
                v-bind="item"
                :key="`${item.label}-${idx}`"
                :item="item"
                :selected="selectionMenu && selectedItem?.value === item.value"
                :selection-menu-child="selectionMenu"
                @change="handleSelection"
              />
            </slot>
          </ul>
        </template>
      </KPop>
    </KToggle>
  </div>
</template>

<script lang="ts" setup>
import type { PropType, Ref } from 'vue'
import { computed, onMounted, ref, watch, nextTick } from 'vue'
import type { ButtonAppearance, DropdownItem, PopPlacements } from '@/types'
import { ButtonAppearances } from '@/types'
import KButton from '@/components/KButton/KButton.vue'
import KTooltip from '@/components/KTooltip/KTooltip.vue'
import KPop from '@/components/KPop/KPop.vue'
import KToggle from '@/components/KToggle'
import KDropdownItem from './KDropdownItem.vue'
import { ChevronDownIcon } from '@kong/icons'

const props = defineProps({
  selectionMenu: {
    type: Boolean,
    default: false,
  },
  appearance: {
    type: String as PropType<ButtonAppearance>,
    default: 'primary',
    validator: (value: ButtonAppearance) => {
      // @ts-ignore
      if (value === 'menu' || value === 'selectionMenu') {
        console.warn('KDropdown: the usage for the `appearance` prop has changed. Please see the migration guide for more details: https://alpha--kongponents.netlify.app/guide/migrating-to-version-9.html#kdropdownmenu')
      }

      return Object.values(ButtonAppearances).includes(value)
    },
  },
  triggerText: {
    type: String,
    default: '',
  },
  /**
   * @deprecated
   * TODO: [beta] remove this prop
   */
  icon: {
    type: String,
    default: '',
    validator: (value: string): boolean => {
      if (value) {
        console.warn('KDropdown: `icon` prop is deprecated. Please slot an icon into the `default` slot instead. See KButton docs for more details: https://alpha--kongponents.netlify.app/components/button.html#default')
      }

      return true
    },
  },
  showCaret: {
    type: Boolean,
    default: false,
  },
  width: {
    type: String,
    default: '',
  },
  /**
   * kpopAttributes is used to pass properties directly to the wrapped KPop component
   * Commonly-overridden properties include:
   * - placement
   * - popoverClasses
   * - target
   */
  kpopAttributes: {
    type: Object,
    default: null,
  },
  items: {
    type: Array as PropType<Array<DropdownItem>>,
    default: () => [],
    validator: (items: DropdownItem[]) => !items.length || items.every(i => i.label !== undefined),
  },
  disabled: {
    type: Boolean,
    default: false,
  },
  disabledTooltip: {
    type: String,
    default: '',
  },
  /**
   * @deprecated in favor of the "triggerText" prop
   */
  label: {
    type: String,
    default: '',
    validator: (value: string) => {
      if (value) {
        console.warn('KDropdown: `label` prop is deprecated. Please use `triggerText` prop instead. See the migration guide for more details: https://alpha--kongponents.netlify.app/guide/migrating-to-version-9.html#kdropdownmenu')
      }

      return true
    },
  },
})

const emit = defineEmits<{
  (e: 'toggleDropdown', value: boolean): void;
  (e: 'change', value: DropdownItem): void;
}>()

const hidePopover = ref<boolean>(false)

const tooltipComponent = computed(() => props.disabledTooltip ? KTooltip : 'div')

const defaultKPopAttributes = {
  hideCaret: true,
  popoverClasses: 'k-dropdown-popover',
  popoverTimeout: 0,
  positionFixed: true,
  placement: 'bottomStart' as PopPlacements,
}

const boundKPopAttributes = {
  ...defaultKPopAttributes,
  ...props.kpopAttributes,
  width: props.width ? props.width : undefined,
  popoverClasses: `${defaultKPopAttributes.popoverClasses} ${props.kpopAttributes?.popoverClasses || ''}`,
}

const triggerButtonText = computed((): string => selectedItem.value?.label || props.triggerText || props.label)

const selectedItem = ref<DropdownItem>()

const handleSelection = (item: DropdownItem): void => {
  if (!props.selectionMenu) {
    return
  }

  selectedItem.value = item
}

const handleCloseDropdown = async (): Promise<void> => {
  hidePopover.value = true

  // reset the hidePopover value so it's ready for the next time the dropdown is opened
  // need nextTick to ensure the popover is hidden before resetting the value
  await nextTick(() => {
    hidePopover.value = false
  })
}

const handleTriggerToggle = (isToggled: Ref<boolean>, toggle: () => void, isOpen: boolean): boolean => {
  // avoid toggling twice for the same event
  if (isToggled.value !== isOpen) {
    toggle()
    emit('toggleDropdown', isToggled.value)
  }

  return isToggled.value
}

watch(selectedItem, (newVal, oldVal): void => {
  if (newVal && newVal !== oldVal) {
    emit('change', newVal)
  }
})

onMounted(() => {
  if (props.items) {
    const selectionArr = props.items.filter(item => item.selected)

    if (selectionArr.length) {
      selectedItem.value = selectionArr[0]
    }
  }
})
</script>

<style lang="scss" scoped>
/* Component styles */

.k-dropdown {
  width: fit-content;

  :deep(.k-popover.k-dropdown-popover) {
    border: var(--kui-border-width-10, $kui-border-width-10) solid var(--kui-color-border, $kui-color-border);
    border-radius: var(--kui-border-radius-30, $kui-border-radius-30);
    margin-top: var(--kui-space-30, $kui-space-30);
    padding: var(--kui-space-0, $kui-space-0);

    ul {
      margin: 0;
      padding: var(--kui-space-20, $kui-space-20) 0;
    }
  }
}
</style>
