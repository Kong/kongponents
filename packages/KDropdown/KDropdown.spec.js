import { mount } from '@vue/test-utils'
import { default as KDropdown, alignments } from './KDropdown.vue'

const Alignments = Object.keys(alignments)

let rendersAlignments = (position) => {
  it(`renders with ${position} class`, () => {
    const wrapper = mount(KDropdown, {
      propsData: {
        'position': position
      }
    })

    expect(wrapper.find('.k-dropdown-menu').html()).toEqual(expect.stringContaining(`k-dropdown-menu-${position}`))
  })
}

describe('KDropdown', () => {
  Alignments.map(position => {
    rendersAlignments(position)
  })

  it('correctly adds active class', () => {
    const wrapper = mount(KDropdown, {
      propsData: {
        items: [
          { name: 'Item 1', selected: true },
          { name: 'Item 2', selected: false }
        ]
      }
    })

    expect(wrapper.find('.k-dropdown-menu > li:first-child').classes('is-selected')).toBe(true)
  })

  it('emits proper item and index on click', async () => {
    const wrapper = mount(KDropdown, {
      propsData: {
        items: [
          { name: 'Item 1', selected: true },
          { name: 'Item 2', selected: false }
        ]
      }
    })

    wrapper.find('.k-dropdown > .k-button').trigger('click')
    wrapper.find('.k-dropdown-menu > li:last-child').trigger('click')

    expect(wrapper.emitted().selected).toBeTruthy()
    expect(wrapper.emitted().selected[0][0]).toEqual({ name: 'Item 2', selected: true })
    expect(wrapper.emitted().selected[0][1]).toEqual(1)
  })

  it('matches snapshot', () => {
    const wrapper = mount(KDropdown)

    expect(wrapper.html()).toMatchSnapshot()
  })
})
