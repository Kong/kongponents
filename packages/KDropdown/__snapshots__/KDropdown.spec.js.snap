// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`KDropdown matches snapshot 1`] = `
<div class="k-dropdown">
  <!----> <button type="button" class="k-button k-dropdown-toggle secondary" aria-haspopup="true"><span></span></button>
  <ul role="listbox" aria-labelledby="dropdown" class="k-dropdown-menu k-dropdown-menu-left"></ul>
</div>
`;
